src/Runtime/Distributions/Automata/Automaton.State.cs

```diff
@@ -73,73 +59,85 @@ public State(int index, IEnumerable<Transition> transitions, Weight endWeight)
                 : this()
             {
                 Argument.CheckIfInRange(index >= 0, "index", "State index must be non-negative.");
-                Argument.CheckIfNotNull(transitions, "transitions");
-
                 this.Index = index;
-                this.EndWeight = endWeight;
-
-                foreach (var transition in transitions)
-                {
-                    this.DoAddTransition(transition);
-                }
+                this.Data = new StateData(transitions, endWeight);
             }
 
             /// <summary>
-            /// Gets the automaton which owns the state.
+            /// Returns where this State represents some valid state in automaton.
             /// </summary>
-            /// <remarks>
-            /// Owner is not serialized to avoid circular references. It has to be restored manually upon deserialization.
-            /// BinaryFormatter and Newtonsoft.Json handle circular references differently by default.
-            /// At the same time [DataMember] does the right thing, because IsReference=true property on State DataContract
-            /// makes DataContractSerializer handle circular references just fine.
-            /// </remarks>
-            [DataMember]
-            [NonSerializedProperty]
-            public TThis Owner { get; internal set; }
+            public bool IsNull => this.Data == null;
 
             /// <summary>
-            /// Helper method for Newtonsoft.Json to skip serialization of <see cref="Owner"/> property.
+            /// Automaton to which this state belongs.
             /// </summary>
-            public bool ShouldSerializeOwner() => false;
+            public Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TThis> Owner { get; }
 
             /// <summary>
             /// Gets the index of the state.
             /// </summary>
-            [DataMember]
-            public int Index { get; internal set; } // TODO: setter of this property is needed only for the state removal procedure
+            public int Index { get; }
 
             /// <summary>
             /// Gets or sets the ending weight of the state.
             /// </summary>
-            [DataMember]
-            public Weight EndWeight { get; set; }
-            
+            /// <remarks>
+            /// C# compiler disallows to use property setter if it sees that <see cref="States"/> instance is a temporary.
+            /// It is not smart enough to understand that property setter actually changes something behind a reference.
+            /// To overcome this issue special <see cref="SetEndWeight"/> method is added calling which is equivalent
+            /// to calling property setter but is not rejected by compiler.
+            /// </remarks>
+            public Weight EndWeight
+            {
+                get => this.Data.EndWeight;
+                set => this.Data.EndWeight = value;
```

Comment:
Remove the setter then? This is more confusing than useful if it cannot be used.